### PR TITLE
adding initial mailer protocol for subsequent implementations

### PR DESCRIPTION
**Pull request by LoganWright**
```swift
let email = Email(
    from: "me@foo.com",
    to: "you@you.bar",
    subject: "Just sayin' hi",
    body: "Hello from mailer"
)

try drop.mailer.send(email)
```

If we like the style, I'll make do some file cleanup and merge it

### CODE REVIEW

File: Sources/Vapor/Droplet/Droplet.swift [hunk 1]
```diff
@@ -97,6 +97,8 @@ public class Droplet {
     /// Storage to add/manage dependencies, identified by a string
     public var storage: [String: Any]
 
+    /// Implemented by your email client
+    public var mailer: Mailer
```
Comment (tanner0101) [inline Sources/Vapor/Droplet/Droplet.swift]: Using just `mail` seems nice here.

```swift
drop.mail.send(...)
```

File: Sources/Vapor/Mail/Mailer.swift [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        try client.send(batch, using: credentials)
+    }
+
+    private func makeClient() throws -> SMTPClient<TCPClientStream> {
+        return try SMTPClient(host: host, port: port, securityLayer: securityLayer)
+    }
+}
+
+extension SMTPMailer {
+    /*
```
Comment (tanner0101) [inline Sources/Vapor/Mail/Mailer.swift]: `///` comments

File: Sources/Vapor/Mail/Mailer.swift [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            credentials: credentials
+        )
+    }
+}
+
+/// To avoid forcing users to deal with an optional on mailer 
+/// which would be annoying long term, this is a place holder
+/// that simply throws, but provides info on how to setup
+/// a proper mailer in the error
+final class UnimplementedMailer: Mailer {
```
Comment (tanner0101) [inline Sources/Vapor/Mail/Mailer.swift]: We use `try drop.mail()` syntax in other places to achieve this.

If optional is not set, it throws some "not configured" error

File: Sources/Vapor/Mail/Mailer.swift [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    func send(batch: [Email]) throws
+}
+
+extension Mailer {
+    public func send(batch: [Email]) throws {
+        try batch.forEach(send)
+    }
+}
+
+public final class SMTPMailer: Mailer {
```
Comment (tanner0101) [inline Sources/Vapor/Mail/Mailer.swift]: This should conform to `ConfigInitializable` using an `smtp.json` or something.

File: Sources/Vapor/Mail/Mailer.swift (hunk 3 above)
Comment (LoganWright) [inline Sources/Vapor/Mail/Mailer.swift]: @tanner0101 where do we use that and how does a user set it, then we have to maintain a setter and getter function with a separate variable. This seems like an easier way to maintain this special scenario.

**Pull request merged**